onnx/defs/math/old.cc

```diff
@@ -1294,5 +1295,116 @@ ONNX_OPERATOR_SET_SCHEMA(
           result_shape.mutable_dim(static_cast<int>(axis))->set_dim_value(k);
           updateOutputShape(ctx, 0, result_shape);
           updateOutputShape(ctx, 1, result_shape);
-        }));		
+        }));
+
+static const char* TopK_ver10_doc = R"DOC(
+Retrieve the top-K elements along a specified axis. Given an input tensor of
+shape [a_1, a_2, ..., a_n, r] and integer argument k, return two outputs:
+  -Value tensor of shape [a_1, a_2, ..., a_{axis-1}, k, a_{axis+1}, ... a_n]
+    which contains the values of the top k elements along the specified axis
+  -Index tensor of shape [a_1, a_2, ..., a_{axis-1}, k, a_{axis+1}, ... a_n] which
+   contains the indices of the top k elements (original indices from the input
+   tensor).
+   
+Given two equivalent values, this operator uses the indices along the axis  as
+ a tiebreaker. That is, the element with the lower index will appear first.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    TopK,
+    10,
+    OpSchema()
+        .SetDoc(TopK_ver10_doc)
+        .Input(0, "X", "Tensor of shape [a_1, a_2, ..., a_n, r]", "T")
+        .Input(
+            1,
+            "K",
+            "A 1-D tensor containing a single positive value corresponding to the number of top elements to retrieve",
+            "tensor(int64)")
+        .Output(
+            0,
+            "Values",
+            "Tensor of shape [a_1, a_2, ..., a_{axis-1}, k, a_{axis+1}, ... a_n] "
+            "containing top K values from the input tensor",
+            "T")
+        .Output(
+            1,
+            "Indices",
+            "Tensor of shape [a_1, a_2, ..., a_{axis-1}, k, a_{axis+1}, ... a_n] "
+            "containing the corresponding input tensor indices for the top K "
+            "values.",
+            "I")
+        .TypeConstraint(
+            "T",
+            {"tensor(float16)", "tensor(float)", "tensor(double)"},
+            "Constrain input and output types to float tensors.")
+        .TypeConstraint(
+            "I",
+            {"tensor(int64)"},
+            "Constrain index tensor to int64")
+        .Attr(
+            "axis",
+            "Dimension on which to do the sort.",
+            AttributeProto::INT,
+            static_cast<int64_t>(-1))
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
```

Comment:
Do you have shape inference tests?

Reply:
One comment/question about the shape-inference logic (I understand it was pre-existing shape-inference logic, not part of this PR): this checks that the "axis" input is a 1-dimensional tensor. However, it is usually recommended that scalars be represented as 0-dimensional tensor. This will likely cause some unexpected error later on. I think it is better that we allow 0-dimensional tensors also here.

Reply:
Yes it is the preexisting shape inference code. Spec update in this PR does not change shape logic.
Regarding axis type, because it is an attribute instead of an input, it shall be a int64_t.